#!/usr/bin/python3
# When running the readDArTag function in polyRAD, all the allele sequences are trimmed to the same length (the length of the shortest allele sequence).
# To avoid additional issues during hapID assignment to the vcf generated by polyRAD, this script trims all microhaplotype sequences in a MADC file to the same length specified by the user.


def trim_microhap_seq_to_same_length(madc, microhap_len):
    inp = open(madc, 'r')
    outp = open(madc.replace('.csv', f'_trimmed{microhap_len}bp.csv'), 'w')
    line = inp.readline()  # Skip the first line (header)
    outp.write(line)  # Write header to output file
    line = inp.readline()  # Read the first sequence line
    while line:
        # Get original sequence and its reverse complement
        line_array = line.strip().split(',')
        trimmed =line_array[2].upper()[:microhap_len]
        outp.write(','.join([line_array[0], line_array[1], trimmed] + line_array[3:]) + '\n')
        line = inp.readline()
    inp.close()
    outp.close()



if __name__=='__main__':

    import argparse

    parser=argparse.ArgumentParser(description="Generate stats from missing allele count")

    parser.add_argument('MADC',
                        help='MADC file with fixed allele IDs and sequences')

    parser.add_argument('microhap_len', type=int,
                        help='Length of microhaplotype sequences to trim to')

    args=parser.parse_args()

    trim_microhap_seq_to_same_length(args.MADC, args.microhap_len)
