#!/usr/bin/python3


def get_botloci(botloci):
    inp = open(botloci)
    line = inp.readline()
    botloci_dict = {}
    cnt = 0
    while line:
        line_array = line.strip().split('_')
        cnt += 1
        if line_array[0] in botloci_dict:
            botloci_dict[line_array[0]].append(int(line_array[1]))
        else:
            botloci_dict[line_array[0]] = [int(line_array[1])]
        line = inp.readline()
    inp.close()
    print('Marker loci on bottom strand:', cnt)
    return(botloci_dict)


def rev_complement(seq):
    complement = str.maketrans("ACTG", "TGAC")
    rev_com = seq.translate(complement)[::-1]
    return(rev_com)


def get_hap_seq_from_vcf(vcf, botloci_dict):
    # polyRAD outputs all haplotype sequences based on the reference genome
    # this means it generates reverse complement of the haplotypes amplified from the bottom strand based on DArTag design
    # Need to convert those back to the sequences matching the DArTag MADC haplotypes in order to check haplotype status
    inp = open(vcf)
    outp = open(vcf.replace('.vcf', '_hap.fa'), 'w')
    line = inp.readline()
    cnt = 0
    while line:
        if not line.startswith('#'):
            # chr1.1	532535	.	CAACGGAACATATAAAGATATCCACTTCTCT
            line_array = line.strip().split('\t')
            bottom = 'false'
            if line_array[0] in botloci_dict:
                for pos in botloci_dict[line_array[0]]:
                    if abs(pos - int(line_array[1])) <= 81:
                        bottom = 'true'
                        cnt += 1
                    else:
                        pass
            else:
                pass
            hap_array = [line_array[3]] + line_array[4].split(',')
            index = 0
            while index < len(hap_array):
                if bottom == 'false':
                    outp.write('>' + line_array[0] + '_' + line_array[1].zfill(9) + '_' + str(index + 1).zfill(3) + '\n')
                    outp.write(hap_array[index] + '\n')
                else:
                    outp.write('>' + line_array[0] + '_' + line_array[1].zfill(9) + '_' + str(index + 1).zfill(3) + '\n')
                    outp.write(rev_complement(hap_array[index]) + '\n')
                index += 1
        line = inp.readline()
    print('Marker loci made reverse complement:', cnt)
    inp.close()
    outp.close()


if __name__=='__main__':

    import argparse

    parser=argparse.ArgumentParser(description="Extract a subset of SNPs from gff3")

    parser.add_argument('botloci', help='Marker loci from the bottom strand')

    parser.add_argument('vcf',
                        help='vcf generated by running ReadDArTag in polyRAD')

    args=parser.parse_args()

    botloci_dict = get_botloci(args.botloci)
    
    get_hap_seq_from_vcf(args.vcf, botloci_dict)
